### ISSUE TITLE
Map rotation?

### ISSUE THREAD

**Issue by Pekin2Paris** [moagrius/TileView#239]
Many thanks to authors about this work. I found map rotation is very useful when we check details of the map (e.g. in indoor space, sometime when you go forward, you need to imagine you are walking backward without map rotation). I am poor in coding, but I'll have a try. Is this function easy to be implemented?

**Comment by osmancoskuner** [comment]
Map rotation is also the main request I get from the people whom I showed my app using TileView. I have been using your library for a long long time.. I really like what you are doing and the quality of your code is great. However, without that feature I don't think I can create an app that will be accepted and used by the majority of users so I believe rotation is vital for any map viewer.

You mentioned a couple of times that integrating rotation feature will require major code rewriting.. I was hopping v2.0 will come with this feature :( Is there any way we can convince you to rethink that decision? I know it is huge work. Maybe we can help you code it or maybe we can start a kickstarter project or something like that to support you..

**Comment by moagrius** [comment]
Rotating the actual tiled image wouldn't be terribly difficult or conflict with existing logic; but virtually everything else would.  The biggest single problem with rotation is that we're have rectangular bounds.  You really need an _infinite plane_ to be able to freely move within that space.  Instead, imagine we've got a single sheet of paper behind the glass of your screen.  Play with that scenario for a few minutes and it quickly becomes apparent why rotation wouldn't work well.  Imagine a user starts off at top left and immediately starts to rotate - we'd either have to show un-rendered space (which so far has been prohibited) or limit rotation to positions where the rotated aspect would be filled with rendered tiles.  This would make it very limited in usability.  So, just by virtue of the rectangular nature of the image, rotation becomes so problematic as to become prohibitive.  Beyond that, it makes virtually every other piece of functionality _worse_ - where would scrolling stop?  It'd no longer be totalWidth - viewport.width, since a rotated edge would almost always be angled, and not a clean line that matched the viewport.  Similarly, tile intersection right now is _really, really_ fast because we can just divide the total width by the tile size grab ever row and column between - with rotation, that goes out the window as well.

Bottom line is that anyone who wants to _try_ to do rotation is certainly welcome to, but without addressing the fundamental issues of rectangular bounds, I don't see how it'll work - I certainly will not be implementing rotation unless I can find away to work around the limits described above.

Hope that all makes sense.

**Comment by peterLaurence** [comment]
Sure implementing map rotation would have a performance cost. I developed an application with the same functionality as TileView, but in a different langage (Python + Kivy), and I agree with the infinite plane. But there were no "bounds" : the user was able to scroll out of rendered space. To address the issue when user is lost in an un-rendered space, i had a button to center the view on the current position (the app used GPS position).
As for intersection, we would have to take into account the angle of rotation. For example, we have screen with an height H, and a width W. The map is rotated by an angle α. To calculate tile intersection, i would just use the same method as it is implemented right now in TileView, but with a "projected" rectangle with dimensions :
H' = H.cos(α) + W.sin(α)
W' = H.sin(α) + W.cos(α)

That's just ideas i have right now, but i haven't started yet to think about it seriously.

**Comment by moagrius** [comment]
Yeah, I'm not sure the benefits or rotation would justify modifications required - as I've said, if anyone wants to take a stab at it, feel free, but I don't see myself making the attempt anytime soon; probably ever.

**Comment by setoba1192** [comment]
I'm trying to tile image, but it create a file with 0kb. Can you help me? 

I'm running this command:

c:\path\to\imagemagick\convert c:\path\to\big\image.png -crop 256x256 -set filename:tile "%%[fx:page.x/256]_%%[fx:page.y/256]" +repage +adjoin c:\path\to\tile\folder\tile-%%[filename:tile].png"

**Comment by alanlee2610** [comment]
hi moagrius, let's say if i have a square map and i zoom out to have the overview of the map ( to render all the tile so that there is no unrendered tile), is that possible for me to achieve the rotation correctly?

**Comment by moagrius** [comment]
i'm not sure there's a "yes" or "no" answer.  it depends on what you're doing and what you expect.  the pixels of the map would rotate, but so would the markers (which you probably don't want).  any other views hit areas would not rotate either, and gestures would be similarly flipped.  if none of that matters, and you just want to show the entire image flipped, then i suppose it might "work".

tl;dr rotation using android methods will just move pixels, and full rotation is not provided or supported by this library

**Comment by Pekin2Paris** [comment]
@alanlee2610 I recently had a look at [OSMDroid](https://github.com/osmdroid/osmdroid). It supports global rotation including tile images, markers, and touch events, but it may be a lot of work on development.

**Comment by dilara-pointr** [comment]
Hello! I am trying to tackle the problem of zooming and dragging while rotating. We previously enabled rotation while limiting the dragging and zooming. We increased the viewport padding to prevent blur tiles. However, in this stage of our product we want to open zoom and drag interactions. What I am doing is similar to @peterLaurence 's solution. I recompute the constraints according to the pivot point and the rotation angle. This doesn't always give good results, especially when zooming. Also I haven't handled the computing the viewport and rendering yet. I would love to discuss possible solutions and get some pointers from you @peterLaurence

**Comment by peterLaurence** [comment]
@dilara-pointr indeed TileView wasn't designed with a rotatable viewport in mind. It sure is a challenge as it requires some additional logic to get the visible tiles of a rotated viewport. This is one of the reasons this feature is not implemented. But if you got something (almost) working, please share or link to your branch so we can improve it.
And who knows, maybe it could be merged upstream :)

**Comment by dilara-pointr** [comment]
Hi! I have implemented rotation feature with drag and zoom properties. Haven't pushed the changes yet. Will adjust the code to the origin and let you guys know the branch.

**Comment by dilara91** [comment]
Hi. This is my personal account. I have added my PR for this problem. @moagrius @peterLaurence 
I hope it will be useful for you. I am not sure if it was appropriate to open PR. I was not sure. If not please remove. My fork is here -> https://github.com/dilara91/TileView/tree/feature/rotatable-map
